src/inc/til/spsc.h

```diff
@@ -0,0 +1,526 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+// til::spsc::details::arc requires std::atomic<size_type>::wait() and ::notify_one() and at the time of writing no
+// STL supports these. Since both Windows and Linux offer a Futex implementation we can easily implement this though.
+// On other platforms we fall back to using a std::condition_variable. We prefer our own Windows/Linux implementations
+// over a potential C++20 one, due to our implementation already being "optimal" on those platforms.
+// For instance the atomic "size_type" is 32 Bit, which happens to be the exact size Linux futex
+// supports, allowing us to leverage that futex without any indirections like hash tables.
+#if defined(_WIN32_WINNT) && _WIN32_WINNT >= _WIN32_WINNT_WIN8
+#define _TIL_SPSC_DETAIL_POSITION_IMPL_WIN 1
+#elif __linux__
+#define _TIL_SPSC_DETAIL_POSITION_IMPL_LINUX 1
+#elif __cpp_lib_atomic_wait >= 201907
+#define _TIL_SPSC_DETAIL_POSITION_IMPL_NATIVE 1
+#else
+#define _TIL_SPSC_DETAIL_POSITION_IMPL_FALLBACK 1
+#endif
+
+namespace til::spsc // Terminal Implementation Library. Also: "Today I Learned"
+{
+    using size_type = uint32_t;
+
+    namespace details
+    {
+        static constexpr size_type position_mask = std::numeric_limits<size_type>::max() >> 2u; // 0b00111....
+        static constexpr size_type revolution_flag = 1u << (std::numeric_limits<size_type>::digits - 2u); // 0b01000....
+        static constexpr size_type drop_flag = 1u << (std::numeric_limits<size_type>::digits - 1u); // 0b10000....
+
+        inline void validate_size(size_t v)
+        {
+            if (v > static_cast<size_t>(position_mask))
+            {
+                throw std::overflow_error{ "size too large for spsc" };
+            }
+        }
+
+#if _TIL_SPSC_DETAIL_POSITION_IMPL_NATIVE
+        using atomic_size_type = std::atomic<size_type>;
+#else
+        struct atomic_size_type
+        {
+            size_type load(std::memory_order order) const noexcept
+            {
+                return _value.load(order);
+            }
+
+            void store(size_type desired, std::memory_order order) noexcept
+            {
+#if _TIL_SPSC_DETAIL_POSITION_IMPL_FALLBACK
+                // We must use a lock here to prevent us from modifying the value
+                // in between wait() reading the value and the thread being suspended.
+                std::unique_lock<std::mutex> lock{ _m };

```

Comment:
[`std::condition_variable` works only with `std::unique_lock<std::mutex>`](https://en.cppreference.com/w/cpp/thread/condition_variable/wait), because `wait()` releases the lock before yielding the current thread to the OS and reacquires the lock when the thread is awoken and only a `std::unique_lock` allows the lock to unlocked/locked repeatedly.

Reply:
But in this method, we don't use `std::condition_variable::wait`. It's below, in the `wait` method that we use `std::condition_variable::wait`.

Reply:
Oh right, thanks! I'll fix that!